src/actions/guides/database/pagination.cr

```diff
@@ -0,0 +1,422 @@
+
+    This module will add the `paginate` method to you actions. This is what
+    you'll use to paginate your query results.
+
+    ## Paginating
+
+    Pagination is easy-peasy, call `paginate` with a query:
+
+    ```crystal
+    class Users::Index < BrowserAction
+      get "/users" do
+        pages, users = paginate(UserQuery.new)
+        html IndexPages, users: users, pages: pages
+      end
+    end
+    ```
+
+    In this case `pages` is a `Lucky::Paginator` with metadata about the
+    current page, total numbers of pages, and helpers for generating paths to
+    next and previous pages.
+
+    The other variable, `users`, is the passed in query (`UserQuery.new`)
+    with a `limit` and `offset` applied so it returns records for just
+    the requested page.
+
+    `paginate` will use the `page` param or fallback to page `1` if `page` is
+    not present. The `page` param can be a query param, a body param, or a
+    route param.
+
+    Next we'll go over how to customize the defaults.
+
+    ## Customizing defaults
+
+    By default `paginate` returns `25` records per page and uses the `page`
+    param to determine which page is requested. But if you want something else,
+    you can!
+
+    ### Customizing records per page
+
+    There are 2 ways to customize items per page.
+
+    You can pass it to the `paginate` method:
+
+    ```crystal
+    paginate(UserQuery.new, per_page: 50)
+    ```
+
+    Or you can customize it with a method on your action:
+
+    ```crystal
+    abstract class BrowserAction < Lucky::Action
+      include Lucky::Paginator::BackendHelpers
+
+      def paginator_per_page : Int32
+        50 # Return a static value
+
+        # Or allow using a param with a default
+        params.get?(:per_page) || 50
+      end
+    end
+    ```
+
+    As you can see you can customize this however you want. You can use a
+    request headers, static value, param, or whatever custom logic you prefer.
+
+    ### Customize current page
+
+    By default `paginate` uses the `page` param. The param can come from a
+    query param, route param, or body param. If the `page` param is not
+    there, it will fall back to page `1`.
+
+    If this logic is not what you want, you can customize it by creating
+    a `paginator_page` that returns an `Int32`
+
+    ```crystal
+    abstract class ApiAction < Lucky::Action
+      include Lucky::Paginator::BackendHelpers
+
+      def paginator_page : Int32
+        request.headers["Page"]? || 1 # Use a header
+      end
+    end
+    ```
+
+    You can use headers, params, or whatever custom logic you'd like.
+
+    ## HTML components
+
+    With just 2 lines of code you can render pagination links in your HTML pages.
+
+    * Add a `needs pages : Lucky::Paginator`
+    * Mount one of the built-in pagination components
+
+    ```crystal
+    class Users::IndexPage < MainLayout
+      needs users : UserQuery
+      needs pages : Lucky::Paginator # Add this to use the paginator
```

Comment:
Side note, I wonder if we had a `needs_paginated users : UserQuery` that does both of these would be helpful 🤔

Reply:
I thought about it but it is really hard to do with the type system and also makes it a bit confusing because you're never sure if it is paginated or not :(

You could do a generic `Paginator(T)` with a query or whatever but then I think that gets fairly confusing. The other plus side of this is we can use the Paginator for arrays, or any enumberable at some point.

With that said I am open to changing this if we come up with something!